Add forward AD test for op info

albanD opened the pull request:
The new OpInfo flag has the following semantic:
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we check that the corresponding error is raised

All the added tests take 3s to run for CPU builds and 1min for GPU builds which should be pretty negligible compared to the test_ops runtime for each of these arch.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58094 Forward AD formulas batch 3
* #57863 Forward AD formulas batch 2
* #57768 Forward AD formulas batch 1
* #57767 Codegen inplace forward AD formula from out of place one if needed
* **#57701 Add forward AD test for op info**
* #57633 Add forward AD gradcheck

Differential Revision: [D28387767](https://our.internmc.facebook.com/intern/diff/D28387767)

albanD committed:
Add forward AD test for op info

[ghstack-poisoned]

albanD committed:
Update on "Add forward AD test for op info"

The new OpInfo flag has the following semantic:
- If not specified, that means the given function does not have forward AD implemented. And we check that the corresponding error is raised
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we skip the test altogether (this happen right now because some formulas are wrong for complex and because the forward AD code make complex input invalid for some linalg functions)





[ghstack-poisoned]

albanD committed:
Update on "Add forward AD test for op info"

The new OpInfo flag has the following semantic:
- If not specified, that means the given function does not have forward AD implemented. And we check that the corresponding error is raised
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we skip the test altogether (this happen right now because some formulas are wrong for complex and because the forward AD code make complex input invalid for some linalg functions)





[ghstack-poisoned]

albanD committed:
Update on "Add forward AD test for op info"

The new OpInfo flag has the following semantic:
- If not specified, that means the given function does not have forward AD implemented. And we check that the corresponding error is raised
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we skip the test altogether (this happen right now because some formulas are wrong for complex and because the forward AD code make complex input invalid for some linalg functions)





[ghstack-poisoned]

albanD committed:
Update on "Add forward AD test for op info"

The new OpInfo flag has the following semantic:
- If not specified, that means the given function does not have forward AD implemented. And we check that the corresponding error is raised
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we skip the test altogether (this happen right now because some formulas are wrong for complex and because the forward AD code make complex input invalid for some linalg functions)





[ghstack-poisoned]

albanD committed:
Update on "Add forward AD test for op info"

The new OpInfo flag has the following semantic:
- If not specified, that means the given function does not have forward AD implemented. And we check that the corresponding error is raised
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we skip the test altogether (this happen right now because some formulas are wrong for complex and because the forward AD code make complex input invalid for some linalg functions)





[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -221,6 +223,19 @@ def test_inplace_gradgrad(self, device, dtype, op):
+        if op.supports_forward_ad:
+            self._grad_test_helper(device, dtype, op, op.get_op(), check_forward_ad=True)
+        else:
+            err_msg = r"Trying to use forward AD with .* that does not support it\."
```
Really nice test

mruberry reviewed (approved):
This is a brilliant extension of our automated testing

albanD committed:
Update on "Add forward AD test for op info"

The new OpInfo flag has the following semantic:
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we check that the corresponding error is raised

All the added tests take 3s to run for CPU builds and 1min for GPU builds which should be pretty negligible compared to the test_ops runtime for each of these arch.




[ghstack-poisoned]

albanD committed:
Update on "Add forward AD test for op info"

The new OpInfo flag has the following semantic:
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we check that the corresponding error is raised

All the added tests take 3s to run for CPU builds and 1min for GPU builds which should be pretty negligible compared to the test_ops runtime for each of these arch.




[ghstack-poisoned]

albanD committed:
Update on "Add forward AD test for op info"

The new OpInfo flag has the following semantic:
- If it says that it supports forward AD, we run gradcheck with forward AD to ensure it is correct
- If it says that it does not support it, we check that the corresponding error is raised

All the added tests take 3s to run for CPU builds and 1min for GPU builds which should be pretty negligible compared to the test_ops runtime for each of these arch.




[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387767).
